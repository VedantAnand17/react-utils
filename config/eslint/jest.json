{
  "env": {
    "browser": true,
    "jest": true
  },
  "plugins": [
    "jest"
  ],
  "rules": {
    "@typescript-eslint/no-implied-eval": 0,
    "global-require": 0,
    "import/no-dynamic-require": 0,
    "import/no-extraneous-dependencies": 0,
    "no-console": 0,
    "no-new-func": 0,
    "react/require-default-props": 0
  }
}