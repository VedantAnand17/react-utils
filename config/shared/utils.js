/**
 * Misc utilities reusable in different JS-format config files.
 */
/* eslint-disable import/no-extraneous-dependencies */

const fs = require('fs');
const path = require('path');
const cssesc = require('cssesc');

const { interpolateName } = require('loader-utils');

/**
 * Normalizes file path to OS-independent format (adopted from css-loader).
 * @ignore
 * @param {string} file
 * @return {string}
 */
function normalizePath(file) {
  return path.sep === '\\' ? file.replace(/\\/g, '/') : file;
}

// eslint-disable-next-line no-control-regex
const filenameReservedRegex = /[<>:"/\\|?*]/g;
// eslint-disable-next-line no-control-regex
const reControlChars = /[\u0000-\u001f\u0080-\u009f]/g;

function escapeLocalident(localident) {
  return cssesc(
    localident
      // For `[hash]` placeholder
      .replace(/^((-?[0-9])|--)/, '_$1')
      .replace(filenameReservedRegex, '-')
      .replace(reControlChars, '-')
      .replace(/\./g, '-'),
    { isIdentifier: true },
  );
}

/**
 * Returns the name of package containing the folder; i.e. it recursively looks
 * up from the folder for the closest package.json file, and returns the name in
 * that file. It also caches the results from previously fisited folders.
 * @ignore
 * @param {string} folder
 * @return {string}
 */
function getPackageInfo(folder) {
  let res = getPackageInfo.cache[folder];
  if (!res) {
    const p = path.resolve(folder, 'package.json');
    /* eslint-disable global-require, import/no-dynamic-require */
    res = fs.existsSync(p) ? {
      name: require(p).name,
      root: folder,
    } : getPackageInfo(path.resolve(folder, '..'));
    /* eslint-enable global-require, import/no-dynamic-require */
    getPackageInfo.cache[folder] = res;
  }
  return res;
}
getPackageInfo.cache = {};

function getLocalIdent(
  { resourcePath },
  localIdentName,
  localName,
  options = {},
) {
  const packageInfo = getPackageInfo(path.dirname(resourcePath));
  const request = normalizePath(path.relative(packageInfo.root, resourcePath));
  return interpolateName({
    resourcePath,
  }, localIdentName, {
    ...options,
    content: `${packageInfo.name + request}\x00${localName}`,
    context: packageInfo.root,
  }).replace(/\[package\]/gi, packageInfo.name)
    .replace(/\[local\]/gi, localName)
    .replace(/@/g, '-');
}

function generateScopedNameDev(localName, assetPath) {
  // Mind: "css-loader" additionally escapes identifiers returned by
  // getLocalIdent(); the babel plugins - don't, thus we need to escape
  // ourselves.
  return escapeLocalident(
    getLocalIdent(
      { resourcePath: assetPath },
      '[package]___[path][name]___[local]___[hash:base64:6]',
      localName,
      {},
    ),
  );
}

function generateScopedNameProd(localName, assetPath) {
  return escapeLocalident(
    getLocalIdent(
      { resourcePath: assetPath },
      '[hash:base64:6]',
      localName,
      {},
    ),
  );
}

module.exports = {
  generateScopedNameDev,
  generateScopedNameProd,
  getLocalIdent,
};
