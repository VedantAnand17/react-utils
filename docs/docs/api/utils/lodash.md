---
sidebar_position: 0
---

# _ (lodash)
```js
import { _ } from '@dr.pogodin/react-utils`;
```
The `_` is just an alias for [lodash](https://lodash.com/docs) library,
which provides various generic JS helpers.
