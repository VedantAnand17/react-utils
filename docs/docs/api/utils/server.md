# server
```jsx
import { server } from '@dr.pogodin/react-utils';
```
The [server](/docs/api/utils/server) module provides constants and methods
facilitating implementation, setup and launch of a configurable ReactJS web
server.

:::info
Beside holding attached constants and functions documented below, `server`
export itself is also a function, which creates and launches a web-server for
ReactJS application. See [server()] documentation for further details.
:::

**Constants**
- [errors.CODES](#errorscodes) - A map from HTTP status code names to
  the corresponding numeric codes.
- [errors.ERRORS](#errorserrors) - A map from HTTP status code names to
  their pretty-printed names.
- [SCRIPT_LOCATIONS] - The map of valid `location` values in [Script] objects.

**Functions**
- [errors.assert()](#errorsassert) - Validates a value against the given [Joi]
  schema.
- [errors.fail()](#errorsfail) - Throws an error with the given message and
  HTTP status code.
- [errors.getErrorForCode()](#errorsgeterrorforcode) - Given HTTP status code
  returns the corresponding error text.
- [errors.joi](#errorsjoi) - An alias for [Joi] library, which provides methods
  for HTTP request validation.
- [errors.newError()](#errorsnewerror) - Creates a new `Error` object with
  the given error message and attached HTTP status.
- [getDefaultCspSettings()](#getdefaultcspsettings) - Returns a deep copy of
  default [CSP] settings object used by **react-utils**.

## Constants

### errors.CODES
[errors.CODES]: #errorscodes
```tsx
import { server } from '@dr.pogodin/react-utils';

const { CODES } = server.errors;

enum CODES {
  // Same as StatusCodes from 'http-status-codes'.
}
```
Enum of HTTP status codes &mdash; an alias of  **StatusCodes** enum from
the [http-status-codes] library.

**Example**
```jsx
import { server } from '@dr.pogodin/react-utils';

const { CODES } = server.errors;
console.log(CODES.BAD_REQUEST);

// Prints: 400
```

### errors.ERRORS
```jsx
server.errors.ERRORS: object
```
A map from HTTP status code names to their pretty-printed names. It is an
alias for **ReasonPhrases** object from [http-status-codes] library.

**Example**
```jsx
import { server } from '@dr.pogodin/react-utils';

const { ERRORS } = server.errors;
console.log(ERRORS.BAD_REQUEST);

// Prints: Bad Request
```

### SCRIPT_LOCATIONS
```jsx
server.SCRIPT_LOCATIONS: object
```
This object provides the map of valid [Script]'s `location` values:
- `BODY_OPEN` - Right after the opening `<body>` tag.
- `DEFAULT` - In the end of `<body>` tag, just before the main application
  bundle. This is the default location where scripts are injected when provided
  as strings rather than [Script] objects.
- `HEAD_OPEN` - Right after the opening `<head>` tag.

## Functions

### errors.assert()
```jsx
server.errors.assert(value, schema, message = '', statusCode = 400)
```
Validates `value` against the given [Joi] `schema`. Throws an error with
the given error `message` and `statusCode` if the validation fails.

**Arguments**
- `value` - **any** value to validate.
- `schema` - **object** - [Joi] schema.
- `message = ''` - **string** - Optional. Error message. If given, the error
  message from [Joi] will be appended to it, otherwise the error message
  from [Joi] is used as is.
- `statusCode = 400` - **number** - Optional. Error status code. Defaults `400`
  (bad request).

### errors.fail()
```tsx
import { server } from '@dr.pogodin/react-utils';

const { fail } = server.errors;

function fail(
  message: string,
  statusCode: CODES = CODES.INTERNAL_SERVER_ERROR,
): never;
```
Throws an error with the given `message` and HTTP `statusCode`.

**Arguments**
- `message` &mdash; **string** &mdash; Error message.
- `statusCode` - [errors.CODES] - Optional. HTTP status code. Defaults `500`
  (internal server error).

It never **returns** normally.

### errors.getErrorForCode()
```jsx
server.errors.getErrorForCode(code): string
```
Given HTTP status `code` returns the corresponding error text. It is an alias
for **getReasonPhrase()** function from [http-status-codes] library.

**Arguments & Result**
- `code` - **number** - HTTP status code.
- Returns **string** - Error message.

### errors.joi
```jsx
server.errors.joi: object
```
An alias for [Joi] library, which provides methods useful for HTTP request
validation.

**Example**: [Joi] methods are used to create validation schema for the body of
a sample HTTP request.
```jsx
import { server } from '@dr.pogodin/react-utils';

const { joi } = server.errors;
const requestBodySchema = joi.object({
  sampleKey: joi.string().max(16).required(),
});
```

### errors.newError()
```jsx
server.errors.newError(message, statusCode = 500): Error
```
Creates a new `Error` object with given message, and HTTP `statusCode` attached
to the `result` as `result.status` field.

**Arguments & Result**
- `message` - **string** - Error message.
- `statusCode = 500` - **number** - Optional. HTTP status code. Defaults `500`
  (Internal Server Error).
- Returns **Error**.

### getDefaultCspSettings()
```jsx
server.getDefaultCspSettings(): object
```
Returns a deep copy of default [CSP] settings object used by **react-utils**,
with exception of `nonce-xxx` clause in `script-src` directive, which is added
dynamically for each request.

[CSP]: https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP
[ExpressJS]: https://expressjs.com
[helmet]: https://github.com/helmetjs/helmet
[http-status-codes]: https://www.npmjs.com/package/http-status-codes
[Joi]: https://joi.dev/api/?v=17.4.2
[Script]: /docs/api/functions/server#beforerender-script
[server]: /docs/api/utils/server
[SCRIPT_LOCATIONS]: #script_locations
[server()]: /docs/api/functions/server
[Webpack]: https://webpack.js.org
