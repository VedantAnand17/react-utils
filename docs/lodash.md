# Lodash (JS Utilities)

```jsx
import { _ } from '@dr.pogodin/react-utils`;
```

`_` is an alias of [`lodash`](https://lodash.com/docs) library.
