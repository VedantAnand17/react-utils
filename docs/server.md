# Web Server for ReactJS App

```js
import { server } from '@dr.pogodin/react-utils`;
import webpackConfig from 'path/to/your/webpack.config';

server(webpackConfig);

/* Web server is up, running, and serving the app specified by the provided
 * Webpack config */
```

All you need to create and launch a web-server for your ReactJS application.
Allows zero or detailed configuration, supports server-side rendering, and
development tools, including Hot Module Reloading (HMR).

### Reference

- `async server(webpackConfig, [options])` &rArr; `Promise` &rArr;
  `{ expressServer, httpServer }`

  **Arguments**

  - `webpackConfig` (_Object_) &ndash; Webpack configuration used to build
    the frontend bundle. In production mode the server will read out of it
    `context`, `publicPath`, and a few other parameters, necessary to locate
    and serve the app bundle. In development mode the server will use entire
    provided config to build the app bundle in memory, and further watch and
    update it via HMR.

  - `[options]` (_Object_) &ndash; Optional. Additional parameters.

  - `[options.Application]` (_Component_) &ndash; Optional. The root ReactJS
    component of the app to use for the server-side rendering. When not provided
    the server-side rendering is disabled.

  - `options.beforeExpressJsError?: (server) => Promise<boolean>?` &ndash;
    Optional. An async callback to be executed just before the default error
    handler is added to ExpressJS server. If the callback is provided and its
    result resolves to a truly value, `react-utils` won't attach the default
    error handler.

  - `options.beforeExpressJsSetup?: (server) => Promise?` &ndash; Optional.
    An async callback to be executed right after ExpressJS server creation,
    bofore any configuration is performed.

  - `[options.beforeRender]` ([_Before Render Hook_](#before-render-hook))
    &ndash; Optional. The hook to run just before the server-side rendering.
    For each incoming request, it will be executed just before the HTML markup
    is generated at the server. It allows to load and provide the data necessary
    for server-side rendering, and also to inject additional configuration and
    scripts into the generated HTML code.

  - `[options.cspSettingsHook: function]` &ndash; Optional. A hook allowing
    to customize [CSP](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP)
    settings for [helmet](https://github.com/helmetjs/helmet)'s
    `contentSecurityPolicy` middleware on per-request basis.

    If provided it should be a with signature: \
    `(defaultSettings: object, req: object)` &rArr; `object` \
    which gets the default settings (also used without the hook),
    and the incoming request object. The hook response will be passed
    as options to the helmet `contentSecurityPolicy` middleware.

    Currently, the default settings is the following object in production
    environment:
    ```js
    {
      directives: {
        defaultSrc: ["'self'"],
        baseUri: ["'self'"],
        blockAllMixedContent: [],
        fontSrc: ["'self'", 'https:', 'data:'],
        frameAncestors: ["'self'"],
        frameSrc: ["'self'", 'https://*.youtube.com'],
        imgSrc: ["'self'", 'data:'],
        objectSrc: ["'none'"],
        scriptSrc: ["'self'", "'unsafe-eval'", `'nonce-UNIQUE_NONCE_VALUE'`],
        scriptSrcAttr: ["'none'"],
        styleSrc: ["'self'", 'https:', "'unsafe-inline'"],
        upgradeInsecureRequests: [] // Removed in dev mode.
      }
    }
    ```
    It matches the default value used by Helmet with a few updates:
    - YouTube host is whitelisted in the `frameSrc` directive to ensure
      the [`<YouTubeVideo>`](./YouTubeVideo.md) component works.
    - An unique per-request nonce is added to `scriptSrc` directive to
      whitelist auxiliary scripts injected by react-utils. The actual nonce
      value can be fetched by host code via `.cspNonce` field of `req` argument
      of `.beforeRender` hook.
    - `upgradeInsecureRequests` directive is removed in development mode,
      to simplify local testing with http requests.

  - `options.defaultLoggerLogLevel = 'info'` &ndash; Optional. Log level for
    the default logger, which is created if no `logger` option provided.
    Defaults to `info`.

  - `[options.devMode]` (_Boolean_) &ndash; Optional. Pass in `true` to start
    the server in development mode.

  - `[options.favicon]` (_String_) &ndash; Optional. Path to the favicon to use
    by the server. By default no favicon is used.

  - `[options.https]` (_Object_) &ndash; Optional. If provided, HTTPS server
    will be started, instead of HTTP otherwise. The object should provide SSL
    certificate and key via two string fields: `cert`, and `key`.

  - `[options.httpsRedirect]` (_Boolean_) &ndash; Optional. Pass in `true`
    to enable automatic redirection of all incoming HTTP requests to HTTPS.

    To smoothly use it at `localhost` you need to run the server in HTTPS mode,
    and also properly create and install a self-signed SSL sertificate on your
    system. This article is helpful:
    [How to get HTTPS working on your local development environment in 5 minutes](https://medium.freecodecamp.org/how-to-get-https-working-on-your-local-development-environment-in-5-minutes-7af615770eec)

  - `options.logger?: Logger` &ndash; Optional. The logger to use at server
    side. By default [`winston`](https://www.npmjs.com/package/winston) logger
    with console transport is used. The logger you provide, or the default
    `winston` logger otherwise, will be attached to the created ExpressJS server
    object.

  - `options.onExpressJsSetup?: (server) => Promise?` (_Function_) &ndash;
    Optional. An async callback to be triggered when most of the server
    configuration is completed, just before the server-side renderer,
    and the default error handler are attached. You can use it to mount
    custom API routes. The server-side logger can be accessed as `server.logger`.

  - `[port]` (_Number_ or _String_) &ndash; Optional. The port to start
    the server on. Defaults **3000**.

  **Returns** `Promise`resolving to the object with fields

  - `expressServer` (_Object_) &ndash; Created ExpressJS server, with
    the provided, or default logger attached as `expressServer.logger`.

  - `httpServer` (Object) &ndash; Created HTTP(S) server.

- <a name="before-render-hook"></a>
  `async beforeRenderHook(req, config)` &rArr; `Promise` &rArr;
  `{ configToInject, extraScripts, store }`

  **Arguments**

  - `req` (_Object_) &ndash; Incoming ExpressJS HTTP request, with some extra
    fields attached:

    - `.cspNonce: string` &ndash; CSP nonce for `<script>` tags, which should be
      added to the tags injected into the page to allow them to work.

  - `config` (_Object_) &ndash; Application config that server wants to
    inject into generated HTML template.

  **Returns** `Promise` resolving to the object with fields
  
  - `[configToInject]` (_Object_) &ndash; Optional. The actual config object to
    inject into the generated page. If not provided, the `config` provided as
    argument will be injected as is.

  - `[extraScripts]` (_Array of Objects and/or Strings_) &ndash; Optional.
    Additional scripts to inject into the generated page. Each script given
    as a string will be injected as is in the end of generated document's
    `<body>`, just before the main application bundle. Each script given as
    an object is expected to have two fields: `code` holding the actual code
    to inject, and `location` specifying where to inject that `code`. The valid
    locations are exposed via
    the [`server.SCRIPT_LOCATIONS`](#script-locations) object. The scripts
    intended for the same location are injected in the order they are listed
    in the `extraScript` array.

  - `[initialState]` (_Object_) &ndash; Optional. Initial value of the global state.

- <a name="script-locations"></a> `server.SCRIPT_LOCATIONS`
  
  - `server.SCRIPT_LOCATIONS.BODY_OPEN` &ndash; Right after the opening `<body>`
    tag.
  - `server.SCRIPT_LOCATIONS.DEFAULT` &ndash; In the end of `<body>` block, just
    before the main application bundle.
  - `server.SCRIPT_LOCATIONS.HEAD_OPEN` &ndash; Right after the opening `<head>`
    tag.
