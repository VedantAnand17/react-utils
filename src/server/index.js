/**
 * Standard web servers.
 */

import _ from 'lodash';
import http from 'http';
import https from 'https';

/* Polyfill required by ReactJS. */
import 'raf/polyfill';

import serverFactory from './server';
import { SCRIPT_LOCATIONS } from './renderer';

/**
 * Normalizes a port into a number, string, or false.
 * @param {String} value Port name or number.
 * @return Port number (Number), name (String), or false.
 */
function normalizePort(value) {
  const port = _.toNumber(value);
  if (_.isFinite(port)) return port; /* port number */
  if (!_.isNumber(port)) return value; /* named pipe */
  return false;
}

/**
 * Creates and starts a new webserver.
 * @param {Object} webpackConfig The Webpack config that was used to build the
 *  frontend bundle.
 * @param {config} options Additional options:
 *  - Application {Function} - Optional. The root ReactJS component of the app
 *    to use for server-side rendering;
 *  - devMode {Boolean} - Whether the server should be started in dev mode;
 *  - favicon {String} - Path of the favicon to be used by the server;
 *  - logger {Object} - Optional. The logger to use. By default, the console
 *    is used (which is not a good decision performancewise, but it will be
 *    changed soon);
 *  - https {Object} - Optional. If given, HTTPS server will be started with
 *    the specified settings. HTTP server that hanles HTTP > HTTPS redirection
 *    will be also started at the specified port. Expected fields:
 *    - cert {String} - SSL Certificate;
 *    - key {String} - SSL key;
 *  - beforeRender {Function} - The hook into server-side rendering. It will get
 *    incoming request as the argument and it should return a promise that will
 *    resolve to the object with the following fields all optional:
 *    - config {Object} - Config object to be injected into the page;
 *    - extraScripts {Array} - Any additional scripts to be injected into
 *      HTML template;
 *    - store {Object} - Redux store, which state should be used for server-side
 *      rendering, if it is performed, and also injected into HTML template as
 *      the initial state.
 *  - onExpressJsSetup {Function} - Custom setup of ExpressJS server.
 *  - port {String} - Optional. The port to listen (number or name). If not
 *    specified the value will be taken from PORT environmental variable, or
 *    default to 3000 otherwise.
 * @return {Promise} Resolves to the result object has two fields:
 *  - express {Object} - ExpressJS server;
 *  - httpServer {Object} - NodeJS HTTP(S) server.
 */
async function launch(webpackConfig, options) {
  /* Options normalization. */
  const ops = options ? _.clone(options) : {};
  ops.port = normalizePort(ops.port || process.env.PORT || 3000);
  _.defaults(ops, {
    logger: console,
  });

  /* Creates servers, resolves and sets the port. */
  const expressServer = await serverFactory(webpackConfig, ops);

  let httpServer;
  if (ops.https) {
    httpServer = https.createServer({
      cert: ops.https.cert,
      key: ops.https.key,
    }, expressServer);
  } else httpServer = http.createServer(expressServer);

  /* Sets error handler for HTTP(S) server. */
  httpServer.on('error', (error) => {
    if (error.syscall !== 'listen') throw error;
    const bind = _.isString(ops.port) ? `Pipe ${ops.port}` : `Port ${ops.port}`;

    /* Human-readable message for some specific listen errors. */
    switch (error.code) {
      case 'EACCES':
        ops.logger.error(`${bind} requires elevated privileges`);
        process.exit(1);
        break;
      case 'EADDRINUSE':
        ops.logger.error(`${bind} is already in use`);
        process.exit(1);
        break;
      default:
        throw error;
    }
  });

  /* Listening event handler for HTTP(S) server. */
  httpServer.on('listening', () => {
    const addr = httpServer.address();
    const bind = _.isString(addr) ? `pipe ${addr}` : `port ${addr.port}`;
    ops.logger.info(`Server listening on ${bind} in ${
      process.env.NODE_ENV} mode`);
  });

  httpServer.listen(ops.port);

  return {
    expressServer,
    httpServer,
  };
}

launch.SCRIPT_LOCATIONS = SCRIPT_LOCATIONS;

export default launch;
