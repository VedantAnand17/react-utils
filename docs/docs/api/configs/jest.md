# Jest
Default configuration for [Jest] testing framework.

:::caution Beware
It does strong assumptions about the host project's code structure: the host
[Jest] configuration in the `/config/jest` folder of host code; the source code
in `/src` folder; build outputs in `/build` folder; _etc._

For a project with different structure it might be easier to compose
the appropriate Jest config from scratch.
:::

To use this default configuration create two files in the `/config/jest` folder
of host project:

```js title="/config/jest/config.js"
const config = require('@dr.pogodin/react-utils/config/jest/default');

// You can tune config as necessary here.

module.exports = config;
```

```js title="/config/jest/setup.js"
import '@dr.pogodin/react-utils/config/jest/setup';
```

After that point to that host config file from the Jest execution command,
_e.g._:
```bash
jest --config config/jest/config.js
```

[Jest]: https://jestjs.io
