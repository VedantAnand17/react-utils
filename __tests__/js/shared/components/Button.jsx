import { noop, omit } from 'lodash';
import { StaticRouter } from 'react-router-dom/server';

import Button from 'components/Button';
import { snapshot } from 'utils/jest';

const testTheme = {
  active: 'ACTIVE_CLASS',
  button: 'BUTTON_CLASS',
  disabled: 'DISABLED_CLASS',
};

describe('Matches snapshots', () => {
  test('when active', () => {
    snapshot((
      <Button
        active
        onClick={noop}
        theme={testTheme}
      >
        BUTTON
      </Button>
    ));
  });
  test('when disabled', () => {
    snapshot((
      <Button
        disabled
        onClick={noop}
        theme={testTheme}
      >
        BUTTON
      </Button>
    ));
    snapshot((
      <Button
        disabled
        onClick={noop}
        theme={omit(testTheme, 'disabled')}
      >
        BUTTON
      </Button>
    ));
  });
  test('when rendered as link', () => {
    snapshot((
      <StaticRouter>
        <Button
          active
          theme={testTheme}
          to="/SOME/TEST/URL"
        >
          BUTTON
        </Button>
      </StaticRouter>
    ));
  });
});
