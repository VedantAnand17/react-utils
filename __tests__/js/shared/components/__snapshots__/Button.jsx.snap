// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshots when active 1`] = `
<div
  class="-dr-pogodin-react-utils___src-shared-components-Button-style___button___E1FNQT   BUTTON_CLASS -dr-pogodin-react-utils___src-shared-components-Button-style___active___MAe9O6   ACTIVE_CLASS"
  role="button"
  tabindex="0"
>
  BUTTON
</div>
`;

exports[`Matches snapshots when disabled 1`] = `
<div
  class="-dr-pogodin-react-utils___src-shared-components-Button-style___button___E1FNQT   BUTTON_CLASS -dr-pogodin-react-utils___src-shared-components-Button-style___disabled___Br9IWV   DISABLED_CLASS"
>
  BUTTON
</div>
`;

exports[`Matches snapshots when disabled 2`] = `
<div
  class="-dr-pogodin-react-utils___src-shared-components-Button-style___button___E1FNQT   BUTTON_CLASS -dr-pogodin-react-utils___src-shared-components-Button-style___disabled___Br9IWV"
>
  BUTTON
</div>
`;

exports[`Matches snapshots when rendered as link 1`] = `
<a
  class="-dr-pogodin-react-utils___src-shared-components-Button-style___button___E1FNQT   BUTTON_CLASS -dr-pogodin-react-utils___src-shared-components-Button-style___active___MAe9O6   ACTIVE_CLASS"
  href="/SOME/TEST/URL"
>
  BUTTON
</a>
`;
